ace/tao/tao_idl/tao_idl.1
.\" Process this file with
.\" groff -man -Tascii tao_idl.1
.\"
.\" tao_idl.1,v 1.7 1999/11/10 20:38:58 othman Exp
.\"
.TH tao_idl 1 "DECEMBER 1998" "CORBA Tools" "User Manuals"
.SH NAME
tao_idl \- Compile IDL sources for use with the ACE ORB (TAO)
.SH SYNOPSIS
.B tao_idl 
.BI [ OPTION... ]
.I filename
.SH DESCRIPTION
.B tao_idl
is a program that is used to compile
.I IDL
source code.  The IDL compiler generates 6 files from each .idl file, the file
names are obtained by taking the IDL basename and appending C.h, C.i, C.cpp,
S.h, S.i and S.cpp; the client stubs are declared in the C.h file, the
skeletons in the S.h file. Please see the Options list on how to get different
suffixes for these files. 

Both clients and servers should link against the object files for the stubs
and skeletons, this is needed to transparently support collocated objects.

The server skeleton can make use of different lookup strategies to match the
incoming operation with the correct operation at the servant.  Dynamic Hashing
and Perfect Hashing are the two different operation lookup strategies that are
currently supported in tao_idl.  By default, tao_idl tries to use Perfect
Hashing strategy.  If supported and available, gperf generates the lookup
methods for this strategy. 
.SH OPTIONS
.TP
.B \-u
The compiler prints out the options that are given below and exits clean.
.TP
.B \-V
the compiler printouts its version and exits
.TP
.BI "\-Wb," "option_list" 
pass options to the backend
.IP
.BI "skel_export_macro=" "macro_name"

The compiler will emit macro_name right after each class or extern
keyword in the generated skeleton code (S files,) this is needed for
Windows/NT that requires special directives to export symbols from
DLLs, usually the definition is just a space on unix platforms. 
.IP
.BI "skel_export_include=" "include_path"

The compiler will generate code to include include_path at the top of
the generated server header, this is usually a good place to define
the server side export macro.
.IP
.BI "stub_export_macro=" "macro_name"

The compiler will emit macro_name right after each class or extern
keyword in the generated stub code, this is needed for Windows/NT that
requires special directives to export symbols from DLLs, usually the
definition is just a space on unix platforms.
.IP
.BI "stub_export_include=" "include_path"

The compiler will generate code to include include_path at the top of
the client header, this is usually a good place to define the export
macro.
.IP
.BI "export_macro=" "macro_name"

The compiler will emit macro_name right after each class or extern keyword,
this is needed for Windows/NT that requires special directives to export
symbols from DLLs, usually the definition is just a space on unix platforms.
.IP
.BI "export_include=" "include_path"

The compiler will generate code to include include_path at the top of the
client header, this is usually a good place to define the export
macro. 
.IP
.BI "pch_include=" "include_path"

The compiler will generate code to include include_path at the top of
all TAO IDL compiler generated files. This can be used with a
precompiled header mechanism, such as those provided by Borland
C++Builder or MSVC++.
.TP
.B \-E
only invoke the preprocessor
.TP
.BI "-Wp," "option_list"
pass options to the preprocessor
.TP
.BI "\-D" "macro_definition"
passed to the preprocessor
.TP
.BI "\-U" "macro_name"
passed to the preprocessor
.TP
.BI "\-I" "include_path"
passed to the preprocessor
.TP
.BI "\-A" "assertion"
passed to the preprocessor
.TP
.B \-Y
passed to the preprocessor
.TP
.B "\-H perfect_hash"
To specify the IDL compiler to generate skelton code that uses perfect
hashed operation lookup strategy, which is the default
strategy. Perfect hashing uses gperf program, to generate lookup methods.
.TP
.B "\-H dynamic_hash"
To specify the IDL compiler to generate skelton code that uses dynamic
hashed operation lookup strategy.
.TP
.B "\-H binary_search"
To specify the IDL compiler to generate skelton code that uses binary search
based operation lookup strategy.
.TP
.B "\-H linear_search"
To specify the IDL compiler to generate skelton code that uses linear search
based operation lookup strategy.
.TP
.B "\-in"
To generate #include statements with <>'s for the standard include
files (e.g. tao/corba.h) indicating them as non-changing files
.TP
.B "\-ic"
To generate #include statements with ""s for changing standard include
files (e.g. tao/corba.h). 
.TP
.B \-g
To specify the path for the perfect hasing program (GPERF). Default is
$ACE_ROOT/bin/gperf or whatever the ACE_GPERF macro was defined to be
during compilation of TAO.
.TP
.B \-o
To specify the output directory to IDL compiler as to where all the
IDL\-compiler\-generated files are to be put. By default, all the files are
put in the current directory from where
.B tao_idl
is called.
.TP
.B \-hc
Client's header file name ending. Default is "C.h".
.TP
.B \-hs
Server's header file name ending. Default is "S.h".
.TP
.B \-hT
Server's template header file name ending. Default is "S_T.h".
.TP
.B \-cs
Client stub's file name ending. Default is "C.cpp".
.TP
.B \-ci
Client inline file name ending. Default is "C.i".
.TP
.B \-ss
Server skeleton file name ending. Default is "S.cpp".
.TP
.B \-sT
Server template skeleton file name ending. Default is "S_T.cpp".
.TP
.B \-si
Server inline skeleton file name ending. Default is "S_i.cpp".
.TP
.B \-st
Server's template inline file name ending. Default is "S_T.i".
.TP
.B \-t
Temporary directory to be used by the IDL compiler. Default : Resolve
ACE_DEFAULT_TEMP_DIR_ENV. If it is not defined, choose /tmp/.
.TP
.B \-Cw
Output a warning if two identifiers in the same scope differ in
spelling only by case (default for now).
.TP
.B \-Ce
Output an error if two indentifiers in the same scope differ in
spelling only by case. Default output is warning (for now).

Since there is at least one OMG IDL file (sfp.idl, part of the
AVStreams specification) that is inconsistent with its own spelling
rule, we have made a warning the default output until the issue is
resolved.
.TP
.B \-Gc
Generate stubs and skeletons using compiled marshaling.
.TP
.B \-Gi
Generate stubs and skeletons using interpretive marshaling (default for now).
.TP
.BI "\-Ge " "flag"
If the value of
.I flag
is 0,
.B tao_idl
will generate code that will use native C++ exceptions. If the value
of
.I flag
is 1,
.B tao_idl
will generate code that will use environment variable for passing
exceptions.
.TP
.B \-Gl
.B For TAO Core developers only.
Generate locality constrained interface implementation. This flag is
not of general interests and should never be used by end-users.
.TP
.B \-Gp
Generated collocated stubs that use Thru_POA collocation strategy.
(default)
.TP
.B \-Gd
Generated collocated stubs that use Direct collocation strategy.
.TP
.B \-Gt
Generate optimized TypeCodes (unimplemented as yet).
.TP
.B \-Gv
Generate code that supports Object-By-Value.
.TP
.B \-GI
Generate templates files for the servant implementation.
.TP
.BI "\-GIh " "arg"
Servant implemenation header file name ending.
.TP
.BI "\-GIs " "arg"
Servant implemenation skeleton file name ending.
.TP
.BI "\-GIb " "arg"
Prefix to the implementation class names.
.TP
.BI "\-GIe " "arg"
Suffix to the implementation class names.
.TP
.BI "\-GIc " "arg"
Generate copy constructors in the servant implementation template
files.
.TP
.B \-Sa
Suppress generation of the
.I Any
operators
.TP
.B \-Sp
Suppress generation of collocated stubs that use Thru_POA collocation
strategy.
.TP
.B \-Sd
Suppress generation of collocated stubs that use Direct collocation
strategy. (default)
.TP
.B \-St
Suppress generation of the
.I TypeCodes
.SH BUGS
None.
.SH AUTHORS
.IP
Carlos O'Ryan <coryan@cs.wustl.edu>
.IP
Ossama Othman <othman@cs.wustl.edu>
performed man page conversion from original HTML source
.SH "SEE ALSO"
.BR gperf (1)


ace/ChangeLog
Fri Oct 05 05:56:50 2001  Balachandran  <bala@cs.wustl.edu>

        * ACE version 5.2 released.

Thu Oct  4 17:38:21 2001  Steve Huston  <shuston@riverace.com>

        * ace/Synch.h: Added some comments to explain the rationale for
          the different return values from the blocking and timed
          versions of ACE_Null_Semaphore::acquire().  Thanks to Doug
          Schmidt for explaining this.

Thu Oct  4 17:27:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp (recv_log_record): Insure that
          the payload message block is properly aligned for any CDR
          extraction. Many thanks to Bala for noticing this problem and
          sending code for the fix.

Thu Oct  4 17:08:39 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Process_Per_Connection_Logging_Server.cpp:
          In prepare(), call ACE_Process_Options::command_line() rather
          than ACE_Process_Options::process_name() to set the program
          path for the child process.

Thu Oct  4 13:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Handler.cpp: Use a ACE_Message_Block,
          rather than hijacking a CDR block, to read the payload into.

Thu Oct  4 10:14:34 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Donald Acton to the Hall of Fame.

Thu Oct  4 06:39:20 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/config-sunos5.5.h: Added ACE_LACKS_MKSTEMP as SunOS 5.5 does
	  not support mkstemp (). 
	
	* ace/config-sunos5.6.h: Undefined ACE_LACKS_MKSTEMP as Sun 5.6
	  and above support it. 

	  Thanks to Don Hinton <dhinton@swan.fr> for reporting this
	  problem and Chad Elliott for testing this out on a Sun OS 5.5
	  box. 

Wed Oct  3 16:37:41 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (read_1): Reverted this change "Wed Oct  3
	  15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>".  This
	  change is alone not enough to fix the problem that Brad
	  reported. The sample program for the text needs to be fixed to
	  get it working right.

Wed Oct  3 15:05:43 2001  Jeff Parsons <parsons@cs.wustl.edu>

        * ace/CDR_Stream.cpp (read_1);

          Reverted changed made in 

          Fri Aug 17 16:56:16 2001  Jeff Parsons <parsons@cs.wustl.edu>

          to read_1(). A similar change made in several other methods
          is ok, but in this case bytes can be copied into the
          input CDR stream without changing wr_ptr, so the check will
          fail unnecessarily. Thanks to Brad Hoskins <BHoskins@slo.newport.com>
          for pointing this out.

Wed Oct  2 12:18:37 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Server.cpp: Fixed spelling error.
        * examples/C++NPv1/Logging_Handler.cpp: Comment length adjustment.

Tue Oct  2 21:52:31 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Reverted this:
          Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

Tue Oct  2 19:52:57 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* Makefile: Removed some of the files that no longer exist in the
	  repository. The inclusion of these files in the Makefile added
	  noise during the beta cutting process.

Tue Oct 02 15:34:49 2001  Carlos O'Ryan  <coryan@uci.edu>

	* etc/tao_cosevent.doxygen:
	* etc/tao_cosnotification.doxygen:
	  Fixed typos in the tag file names.

Tue Oct  2 15:52:27 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Added Sunny Leung to the Hall of Fame.

Tue Oct  2 15:33:36 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* THANKS: Removed one instance of Peter Phillips
	  <pphillip@opentext.com> from the Hall of Fame.

Tue Oct  2 15:00:37 2001  Balachandran Natarajan  <bala@cs.wustl.edu>

	* ace/CDR_Stream.cpp (clone_from): Copied the GIOP major and minor
	  versions from the incoming CDR stream. Thanks to Sunny Leung 
	  <sleung@opentext.com> for reporting this. This closes bugzilla
	  #1043.  

Tue Oct  2 09:45:39 2001  Steve Huston  <shuston@riverace.com>

        * tests/libDLL_Test.icc:
        * tests/libService_Config_DLL.icc: Add link(exportAll) to the link
          options to export the entrypoints needed for the tests.

Mon Oct  1 20:04:14 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* THANKS: 
	  Added Tushar Nair to the THANKS file.

Mon Oct  1 18:06:48 2001  Priyanka Gontla  <pgontla@ece.uci.edu>

	* ace/Object_Manager.cpp (ACE_UnhandledExceptionFilter):
	  Removed the un-needed #if 0's. Left them before as a safety
	  measure (in case something doesnt work) and forgot to remove
	  them later.

	* ace/Log_Msg.cpp (instance):
	  Fixed the seg faults that are raised when ace is compiled with
	  ACE_NTRACE=0 on SunOS. Added a check before locking and
	  releasing a lock.

	  Thanks to Tushar Nair <ntushar@yahoo.com> for reporting the
	  problem and for providing a fix.

Mon Oct  1 18:27:41 2001  Steve Huston  <shuston@riverace.com>

        * examples/C++NPv1/Logging_Client.cpp: Remove the third (optional)
          arg from getline(). Using a char literal without further
          qualification causes confusion in Unicode builds.

Mon Oct  1 14:44:13 2001  Ossama Othman  <ossama@uci.edu>

	* ace/SSL/SSL_Context.cpp (ssl_library_init):

	  Initialize the locking callbacks in the crypto library before
	  initializing the OpenSSL library.  OpenSSL may require the
	  locking callbacks to be initialized before its own
	  initialization.

	* ace/SSL/SSL_Context.cpp (ssl_library_fini):

	  Perform crypto library resource cleanup before reclaiming the
	  locks used by the locking callbacks.  Crypto library resource
	  cleanup makes use of the locking callbacks.

Mon Oct  1 15:19:11 2001  Douglas C. Schmidt  <schmidt@macarena.cs.wustl.edu>


ace/tao/tao_idl/tao_idl.cpp
// tao_idl.cpp,v 1.20 2001/05/23 19:33:17 parsons Exp

/*

COPYRIGHT

Copyright 1992, 1993, 1994 Sun Microsystems, Inc.  Printed in the United
States of America.  All Rights Reserved.

This product is protected by copyright and distributed under the following
license restricting its use.

The Interface Definition Language Compiler Front End (CFE) is made
available for your use provided that you include this license and copyright
notice on all media and documentation and the software program in which
this product is incorporated in whole or part. You may copy and extend
functionality (but may not remove functionality) of the Interface
Definition Language CFE without charge, but you are not authorized to
license or distribute it to anyone else except as part of a product or
program developed by you or with the express written consent of Sun
Microsystems, Inc. ("Sun").

The names of Sun Microsystems, Inc. and any of its subsidiaries or
affiliates may not be used in advertising or publicity pertaining to
distribution of Interface Definition Language CFE as permitted herein.

This license is effective until terminated by Sun for failure to comply
with this license.  Upon termination, you shall destroy or return all code
and documentation for the Interface Definition Language CFE.

INTERFACE DEFINITION LANGUAGE CFE IS PROVIDED AS IS WITH NO WARRANTIES OF
ANY KIND INCLUDING THE WARRANTIES OF DESIGN, MERCHANTIBILITY AND FITNESS
FOR A PARTICULAR PURPOSE, NONINFRINGEMENT, OR ARISING FROM A COURSE OF
DEALING, USAGE OR TRADE PRACTICE.

INTERFACE DEFINITION LANGUAGE CFE IS PROVIDED WITH NO SUPPORT AND WITHOUT
ANY OBLIGATION ON THE PART OF Sun OR ANY OF ITS SUBSIDIARIES OR AFFILIATES
TO ASSIST IN ITS USE, CORRECTION, MODIFICATION OR ENHANCEMENT.

SUN OR ANY OF ITS SUBSIDIARIES OR AFFILIATES SHALL HAVE NO LIABILITY WITH
RESPECT TO THE INFRINGEMENT OF COPYRIGHTS, TRADE SECRETS OR ANY PATENTS BY
INTERFACE DEFINITION LANGUAGE CFE OR ANY PART THEREOF.

IN NO EVENT WILL SUN OR ANY OF ITS SUBSIDIARIES OR AFFILIATES BE LIABLE FOR
ANY LOST REVENUE OR PROFITS OR OTHER SPECIAL, INDIRECT AND CONSEQUENTIAL
DAMAGES, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.

Use, duplication, or disclosure by the government is subject to
restrictions as set forth in subparagraph (c)(1)(ii) of the Rights in
Technical Data and Computer Software clause at DFARS 252.227-7013 and FAR
52.227-19.

Sun, Sun Microsystems and the Sun logo are trademarks or registered
trademarks of Sun Microsystems, Inc.

SunSoft, Inc.
2550 Garcia Avenue
Mountain View, California  94043

NOTE:

SunOS, SunSoft, Sun, Solaris, Sun Microsystems or the Sun logo are
trademarks or registered trademarks of Sun Microsystems, Inc.

*/

#include "idl.h"
#include "global_extern.h"
#include "fe_extern.h"
#include "drv_private.h"
#include "drv_extern.h"
#include "be.h"
#include "ace/Process.h"
#include "../tao/Version.h"

ACE_RCSID(TAO_IDL, tao_idl, "tao_idl.cpp,v 1.20 2001/05/23 19:33:17 parsons Exp")

#define         IDL_CFE_VERSION        "1.3.0"

// Initialize the BE. The protocol requires only that this routine
// return an instance of AST_Generator (or a subclass thereof).
AST_Generator *
BE_init (void)
{
  tao_cg = TAO_CODEGEN::instance ();

  AST_Generator *g = 0;
  ACE_NEW_RETURN (g,
                  be_generator (),
                  0);

  return g;
}

// Print out a version string for the BE.
void
BE_version (void)
{
  ACE_DEBUG ((LM_DEBUG,
              "%s %s\n",
              ACE_TEXT ("TAO C++ BE, version"),
              ACE_TEXT (TAO_VERSION)));
}

void
DRV_version (void)
{
  ACE_DEBUG ((LM_DEBUG,
              "%s\n"
              "%s %s\n",
              ACE_TEXT (idl_global->prog_name ()),
              ACE_TEXT ("FE: Based on Sun IDL CFE version"),
              ACE_TEXT (IDL_CFE_VERSION)));

  BE_version ();
}

// Fork off a process, wait for it to die.
void
DRV_fork (void)
{
  for (DRV_file_index = 0;
       DRV_file_index < DRV_nfiles;
       ++DRV_file_index)
    {
      ACE_Process_Options options (1, 
                                   TAO_IDL_COMMAND_LINE_BUFFER_SIZE);
      options.creation_flags (ACE_Process_Options::NO_EXEC);
      options.command_line ("%s %s %s", 
                            idl_global->prog_name (), 
                            idl_global->idl_flags (), 
                            DRV_files[DRV_file_index]);
      ACE_Process manager;
      pid_t child_pid = manager.spawn (options);

      if (child_pid == 0)
        {
          // OK, do it to this file (in the child).
          DRV_drive (DRV_files[DRV_file_index]);
          ACE_OS::exit (0);
        }

      if (child_pid == ACE_INVALID_PID)
        {
          ACE_ERROR ((LM_ERROR,
                      ACE_TEXT ("IDL: spawn failed\n")));

          ACE_OS::exit (99);
        }

      // child_pid is the process id of something at this point.
      if (manager.wait () == -1)
        {
          ACE_ERROR ((LM_ERROR,
                      ACE_TEXT ("IDL: wait failed\n")));

          ACE_OS::exit (99);
        }
    }

  // Now the parent process can exit.
  ACE_OS::exit (0);
}

/*
** Drive the compilation
**
** LOGIC:
**
** 1. Initialize the CFE, stage 1. This builds the scope stack
** 2. Initialize the BE. This builds an instance of the generator
** 3. Initialize the CFE, stage 2. This builds the global scope
**    and populates it with the predefined types
** 4. Invoke FE_yyparse
** 5. Check for errors from FE_yyparse. If any, exit now
** 6. Check for undefined forward declared interfaces. If any, exit now
** 7. Check if asked to dump AST. If so, do.
** 8. Invoke BE.
*/

void
DRV_drive (const char *s)
{
  // Set the name of the IDL file we are parsing. This is useful to
  // the backend when it generates C++ headers and files.
  UTL_String *src_file = 0;
  ACE_NEW (src_file,
           UTL_String (s));

  idl_global->idl_src_file (src_file);

  // Pass through CPP.
  if (idl_global->compile_flags () & IDL_CF_INFORMATIVE)
    {
      ACE_DEBUG ((LM_DEBUG,
                  "%s%s %s\n",
                  ACE_TEXT (idl_global->prog_name ()),
                  ACE_TEXT (": preprocessing"),
                  ACE_TEXT (s)));
    }

  DRV_pre_proc (s);

  // Initialize FE stage 1.
  FE_init_stage1 ();

  // Initialize BE.
  AST_Generator *gen = BE_init ();

  if (gen == 0)
    {
      ACE_ERROR ((
          LM_ERROR,
          ACE_TEXT ("IDL: BE init failed to create generator, exiting\n")
        ));

      ACE_OS::exit (99);
    }
  else
    {
      idl_global->set_gen (gen);
    }

  // Initialize FE stage 2.
  FE_init_stage2 ();

  // Parse.
  if (idl_global->compile_flags () & IDL_CF_INFORMATIVE)
    {
      ACE_DEBUG ((LM_DEBUG,
                  "%s%s %s\n",
                  ACE_TEXT (idl_global->prog_name()),
                  ACE_TEXT (": parsing"),
                  ACE_TEXT (s)));
    }

  FE_yyparse ();

  // If there were any errors, stop.
  if (idl_global->err_count () > 0)
    {
      ACE_ERROR ((LM_ERROR,
                  "%s%s %s%s %d %s%s\n",
                  ACE_TEXT (idl_global->prog_name ()),
                  ACE_TEXT (":"),
                  ACE_TEXT (s),
                  ACE_TEXT (": found"),
                  idl_global->err_count (),
                  ACE_TEXT ("error"),
                  ACE_TEXT ((idl_global->err_count () > 1 ? "s" : ""))));

      // Call BE_abort to allow a BE to clean up after itself.
      BE_abort ();

      ACE_OS::exit (ACE_static_cast (int, idl_global->err_count ()));
    }

  // Dump the code.
  if ((idl_global->compile_flags () & IDL_CF_INFORMATIVE)
      && (idl_global->compile_flags () & IDL_CF_DUMP_AST))
    {
      ACE_DEBUG ((LM_DEBUG,
                  "%s%s %s\n",
                  ACE_TEXT (idl_global->prog_name ()),
                  ACE_TEXT (": dump"),
                  ACE_TEXT (s)));
    }

  if (idl_global->compile_flags () & IDL_CF_DUMP_AST)
    {
      ACE_DEBUG ((LM_DEBUG,
                  ACE_TEXT ("Dump of AST:\n")));

      idl_global->root ()->dump (*ACE_DEFAULT_LOG_STREAM);
    }

  // Call the main entry point for the BE.
  if (idl_global->compile_flags () & IDL_CF_INFORMATIVE)
    {
      ACE_DEBUG ((LM_DEBUG,
                  "%s%s %s\n",
                  ACE_TEXT (idl_global->prog_name ()),
                  ACE_TEXT (": BE processing on"),
                  ACE_TEXT (s)));
    }

  BE_produce ();

  // Exit cleanly.
  ACE_OS::exit (0);
}

/*
** LOGIC:
**
** 1. Initialize compiler driver
** 2. Parse command line args
** 3. If more than one file to parse, fork
** 4. Otherwise, for the single file, invoke DRV_drive
*/

int
main (int argc, char *argv[])
{
  // Initialize driver and global variables.
  DRV_init ();

  // Parse arguments.
  DRV_parse_args (argc, argv);

  // If a version message is requested, print it and exit.
  if (idl_global->compile_flags () & IDL_CF_VERSION)
    {
      DRV_version ();
      ACE_OS::exit (0);
    }

  // If a usage message is requested, give it and exit.
  if (idl_global->compile_flags () & IDL_CF_ONLY_USAGE)
    {
      DRV_usage ();
      ACE_OS::exit (0);
    }

  // Fork off a process for each file to process. Fork only if
  // there is more than one file to process.
  if (DRV_nfiles > 1)
    {
      // DRV_fork never returns.
      DRV_fork ();
    }
  else
    {
      // Do the one file we have to parse.
      // Check if stdin and handle file name appropriately.
      if (DRV_nfiles == 0)
        {
          DRV_files[0] = "standard input";
        }

      DRV_file_index = 0;
      DRV_drive (DRV_files[DRV_file_index]);
    }

  ACE_OS::exit (0);

  // NOT REACHED
  return 0;
}
